src/main/java/org/folio/circulation/domain/OverduePeriodCalculatorService.java

```diff
@@ -0,0 +1,105 @@
+package org.folio.circulation.domain;
+
+import org.apache.commons.lang3.ObjectUtils;
+import org.folio.circulation.support.Result;
+import org.joda.time.DateTime;
+import org.joda.time.LocalTime;
+
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+import static java.util.concurrent.CompletableFuture.completedFuture;
+import static org.folio.circulation.support.Result.succeeded;
+import static org.folio.circulation.support.ResultBinding.flatMapResult;
+import static org.joda.time.DateTimeConstants.MINUTES_PER_DAY;
+import static org.joda.time.DateTimeConstants.MINUTES_PER_HOUR;
+import static org.joda.time.Minutes.minutesBetween;
+
+public class OverduePeriodCalculatorService {
+  private static final int ZERO_MINUTES = 0;
+
+  private final CalendarRepository calendarRepository;
+
+  public OverduePeriodCalculatorService(CalendarRepository calendarRepository) {
+    this.calendarRepository = calendarRepository;
+  }
+
+  public CompletableFuture<Result<Integer>> getMinutes(Loan loan, DateTime systemTime) {
+    if (preconditionsAreNotMet(loan, systemTime)) {
+      return completedFuture(succeeded(ZERO_MINUTES));
+    }
+
+    return getOverdueMinutes(loan, systemTime)
+      .thenApply(flatMapResult(om -> adjustOverdueWithGracePeriod(loan, om)));
+  }
+
+  boolean preconditionsAreNotMet(Loan loan, DateTime systemTime) {
+    return loan.isOverdue(systemTime)
+      || loan.getOverdueFinePolicy().getCountPeriodsWhenServicePointIsClosed() == null;
+  }
+
+  CompletableFuture<Result<Integer>> getOverdueMinutes(Loan loan, DateTime systemTime) {
+    DateTime dueDate = loan.getDueDate();
+    boolean countClosed = loan.getOverdueFinePolicy().getCountPeriodsWhenServicePointIsClosed();
+    if (countClosed) {
+      int overdueMinutes = minutesBetween(dueDate, systemTime).getMinutes();
+      return completedFuture(succeeded(overdueMinutes));
+    } else {
+      return calendarRepository
+        .fetchOpeningPeriodsBetweenDates(loan.getCheckoutServicePointId(), dueDate, systemTime, false)
+        .thenApply(r -> r.next(this::getOpeningPeriodsDurationMinutes));
```

Comment:
Maybe extract a method for this, e.g. `minutesOverdueExcludingClosedDays`

These would allow the two branches to be described at the same level of abstraction.

Reply:
done